packages/website/blog/2024-03-25-changes-to-consistent-type-imports-with-decorators.md

```diff
@@ -0,0 +1,101 @@
+---
+authors:
+  - image_url: /img/team/bradzacher.jpg
+    name: Brad Zacher
+    title: typescript-eslint Maintainer
+    url: https://github.com/bradzacher
+description: Changes to consistent-type-imports when used with decorators, experimentalDecorators, and emitDecoratorMetadata
+slug: changes-to-consistent-type-imports-with-decorators
+tags:
+  [
+    consistent-type-imports,
+    experimentalDecorators,
+    emitDecoratorMetadata,
+    typescript-eslint,
+  ]
+title: Changes to consistent-type-imports when used with legacy decorators and decorator metadata
+---
+
+We've made some changes to `consistent-type-imports` to fix some long-standing issues when used alongside `experimentalDecorators: true` and `emitDecoratorMetadata: true`. These changes increase safety and prevent invalid fixes when using decorator metadata.
+
+<!-- truncate -->
+
+## Experimental Decorator Metadata
+
+The `experimentalDecorators` compiler option (referred to as "legacy decorators" from here on) turns on support for an old version of the decorator proposal that was never standardized. Syntactically this proposal isn't too dissimilar to the current [Stage 3 proposal](https://github.com/tc39/proposal-decorators), however they differ in one major way - the use of [metadata reflection](https://rbuckton.github.io/reflect-metadata/) when `emitDecoratorMetadata` is turned on.
+
+When using legacy decorators with decorator metadata and a class is annotated with decorators TypeScript will emit runtime metadata for the class that captures the property types, method parameter types, and method return types. This metadata is incredibly powerful and provides a bridge between the types (which are elided at compile time) and the runtime code. However this runtime code is generated at a cost as it is derived using type information - meaning that the runtime code emitted changes based on the cross-file type information that TypeScript has computed.
+
+To illustrate what this means consider the following snippet:
+
+```ts
+import Foo from 'foo';
+import decorator from 'decorator';
+
+class Clazz {
+  @decorator
+  method(arg: Foo) {}
+}
+```
+
+If the imported name `Foo` resolves to...
+
+- a type then TS will emit metadata that annotates `arg` with one of `Function`, `Object`, `String`, `Number`, or `Boolean` depending on what that type resolves to.
+- an enum then TS will emit metadata that annotates `arg` with one of `String`, `Number`, or `Object` depending on the type of the enum's members.
+- a class declaration:
+  - and the import **_is NOT_** annotated as `import type` then TS will emit metadata that annotates the type of `arg` as `Foo`.
+  - and the import **_IS_** annotated as `import type` then TS will emit metadata that annotates the type of `arg` as `Function`.
+
```

Comment:
I'm of the opinion it's good for narrative flow to end every section with a brief summation of the final point(s) that lead into the next one.

```suggestion

In addition to requiring runtime type information, those metadata emit rules are confusing for developers to reason about.
They necessitate understanding edge cases specific to TypeScript's handling of decorators and type information.

```